### ISSUE TITLE
Kube-proxy "services have not yet been read at least once, cannot construct envvars"

### ISSUE THREAD

**Issue by Pseudow** [kubernetes/kubernetes#110977]
### What happened?

Kube-proxy keep crashing and restarting here are the description of the pod:
```
  Normal   Scheduled       49m                   default-scheduler  Successfully assigned kube-system/kube-proxy-4m5pb to altarise-third
  Warning  Failed          49m (x2 over 49m)     kubelet            Error: services have not yet been read at least once, cannot construct envvars
  Normal   Pulled          45m (x5 over 49m)     kubelet            Container image "k8s.gcr.io/kube-proxy:v1.24.2" already present on machine
  Normal   Created         45m (x3 over 49m)     kubelet            Created container kube-proxy
  Normal   Started         45m (x3 over 49m)     kubelet            Started container kube-proxy
  Normal   SandboxChanged  44m (x3 over 47m)     kubelet            Pod sandbox changed, it will be killed and re-created.
  Normal   Killing         8m21s (x11 over 47m)  kubelet            Stopping container kube-proxy
  Warning  BackOff         4m9s (x159 over 46m)  kubelet            Back-off restarting failed container
```

### How can we reproduce it (as minimally and precisely as possible)?

Having a node behind NAT using wireguard to fix that, setting the wireguard ip in /etc/default/kubelet
```
$ cat /etc/default/kubelet                    
KUBELET_EXTRA_ARGS=--node-ip=10.66.66.3
```
And joining a normal master running correctly.

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.2", GitCommit:"f66044f4361b9f1f96f0053dd46cb7dce5e990a8", GitTreeState:"clean", BuildDate:"2022-06-15T14:22:29Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.2", GitCommit:"f66044f4361b9f1f96f0053dd46cb7dce5e990a8", GitTreeState:"clean", BuildDate:"2022-06-15T14:15:38Z", GoVersion:"go1.18.3", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
$ cat /etc/os-release
Debian GNU/Linux 11 (bullseye)
$ uname -a
Linux altarise-third 5.10.0-15-amd64 #1 SMP Debian 5.10.120-1 (2022-06-09) x86_64 GNU/Linux
```

</details>

### Install tools

<details>
https://kubernetes.io/fr/docs/setup/production-environment/tools/kubeadm/install-kubeadm/
</details>

### Container runtime (CRI) and version (if applicable)

<details>
flannel latest version even though the flannel pod keep restarting because of that
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by dcbw** [comment]
@Pseudow are you able to attach the logs from kube-proxy? The events don't give quite enough detail...

**Comment by Pseudow** [comment]
```
$ cat /var/log/containers/kube-proxy-4cvqh_kube-system_kube-proxy-58562eb99a062ae8df70a54b5b7b3ad1c88ba9f3ec47bfdf8592c9ecabdbdd16.log                                                                  20:36:36 2022-07-07T20:33:52.691802331+02:00 stderr F I0707 18:33:52.691427       1 node.go:163] Successfully retrieved node IP: 10.66.66.3
2022-07-07T20:33:52.691924057+02:00 stderr F I0707 18:33:52.691595       1 server_others.go:138] "Detected node IP" address="10.66.66.3"
2022-07-07T20:33:52.691948045+02:00 stderr F I0707 18:33:52.691653       1 server_others.go:578] "Unknown proxy mode, assuming iptables proxy" proxyMode=""
2022-07-07T20:33:52.734100962+02:00 stderr F I0707 18:33:52.733946       1 server_others.go:206] "Using iptables Proxier"
2022-07-07T20:33:52.734357844+02:00 stderr F I0707 18:33:52.734251       1 server_others.go:213] "kube-proxy running in dual-stack mode" ipFamily=IPv4
2022-07-07T20:33:52.734486549+02:00 stderr F I0707 18:33:52.734407       1 server_others.go:214] "Creating dualStackProxier for iptables"
2022-07-07T20:33:52.734654771+02:00 stderr F I0707 18:33:52.734555       1 server_others.go:501] "Detect-local-mode set to ClusterCIDR, but no IPv6 cluster CIDR defined, , defaulting to no-op detect-local for IPv6"
2022-07-07T20:33:52.734907697+02:00 stderr F I0707 18:33:52.734790       1 proxier.go:259] "Setting route_localnet=1, use nodePortAddresses to filter loopback addresses for NodePorts to skip it https://issues.k8s.io/90259"
2022-07-07T20:33:52.735338582+02:00 stderr F I0707 18:33:52.735169       1 proxier.go:259] "Setting route_localnet=1, use nodePortAddresses to filter loopback addresses for NodePorts to skip it https://issues.k8s.io/90259"
2022-07-07T20:33:52.735896275+02:00 stderr F I0707 18:33:52.735773       1 server.go:661] "Version info" version="v1.24.2"
2022-07-07T20:33:52.735918158+02:00 stderr F I0707 18:33:52.735809       1 server.go:663] "Golang settings" GOGC="" GOMAXPROCS="" GOTRACEBACK=""
2022-07-07T20:33:52.749029222+02:00 stderr F I0707 18:33:52.748824       1 conntrack.go:52] "Setting nf_conntrack_max" nf_conntrack_max=262144
2022-07-07T20:33:52.749485859+02:00 stderr F I0707 18:33:52.749295       1 config.go:444] "Starting node config controller"
2022-07-07T20:33:52.749522947+02:00 stderr F I0707 18:33:52.749337       1 config.go:226] "Starting endpoint slice config controller"
2022-07-07T20:33:52.749536056+02:00 stderr F I0707 18:33:52.749348       1 shared_informer.go:255] Waiting for caches to sync for node config
2022-07-07T20:33:52.749546815+02:00 stderr F I0707 18:33:52.749384       1 shared_informer.go:255] Waiting for caches to sync for endpoint slice config
2022-07-07T20:33:52.749556183+02:00 stderr F I0707 18:33:52.749427       1 config.go:317] "Starting service config controller"
2022-07-07T20:33:52.750362336+02:00 stderr F I0707 18:33:52.750230       1 shared_informer.go:255] Waiting for caches to sync for service config
2022-07-07T20:33:52.849697076+02:00 stderr F I0707 18:33:52.849495       1 shared_informer.go:262] Caches are synced for node config
2022-07-07T20:33:52.849744617+02:00 stderr F I0707 18:33:52.849663       1 shared_informer.go:262] Caches are synced for endpoint slice config
2022-07-07T20:33:52.851014778+02:00 stderr F I0707 18:33:52.850858       1 shared_informer.go:262] Caches are synced for service config
```

but I think it is related to my VPN setup between my nodes. As I can see from this node I cant ping my kubeapiserver I have to specify my wg0 interface which is kind of weird because from my others nodes using the same VPN I can ping kubeapiserver without specifying it. But I dont really understand why it does that.

**Comment by Pseudow** [comment]
I compared my node iptable using the command

```
iptables -t nat -L
```

And it gave me that:
 ![](https://media.discordapp.net/attachments/994312928775897198/994681828155850873/unknown.png?width=1440&height=591)

**Comment by aojea** [comment]
> @Pseudow are you able to attach the logs from kube-proxy? The events don't give quite enough detail...

you need to paste the tail of the logs, there should be a message indicating why kube-proxy can't start

**Comment by Pseudow** [comment]
Actually I found out that it was my node behind NAT that can curl 10.96.0.1:443 (kubeapi server) but in a weird way: sometimes it directly returns an ouput from the api and sometimes it can take up to 40seconds before having a answer. I think this might explain why it keeps crashing. What is actually more weird is that I can easily curl 10.66.66.1:6443 (the kubeapi server on my master node).
![](https://media.discordapp.net/attachments/876549488892870776/995771266390823082/unknown.png?width=688&height=671)
I tried to debug that using tcpdump to see wether it took time from the server to send the answer back or from the client to send it to the master and I found out when I curl 10.96.0.1:443 it directly sends a packet to my master machine which receives it but it doesnt always send a response back.
